okhttp/src/test/java/okhttp3/CacheCorruptionTest.kt

```diff
@@ -0,0 +1,109 @@
+package okhttp3
+
+import okhttp3.internal.io.InMemoryFileSystem
+import okhttp3.mockwebserver.MockResponse
+import okhttp3.mockwebserver.MockWebServer
+import okhttp3.testing.PlatformRule
+import okhttp3.tls.internal.TlsUtil.localhost
+import org.assertj.core.api.Assertions
+import org.junit.After
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import java.io.File
+import java.io.IOException
+import java.net.CookieManager
+import java.net.ResponseCache
+import java.text.DateFormat
+import java.text.SimpleDateFormat
+import java.util.Date
+import java.util.Locale
+import java.util.TimeZone
+import java.util.concurrent.TimeUnit
+import javax.net.ssl.HostnameVerifier
+import javax.net.ssl.SSLSession
+
+class CacheCorruptionTest {
+  @get:Rule
+  var server = MockWebServer()
+
+  @get:Rule
+  var fileSystem = InMemoryFileSystem()
+
+  @get:Rule
+  val clientTestRule = OkHttpClientTestRule()
+
+  @get:Rule
+  val platform = PlatformRule()
+
+  private val handshakeCertificates = localhost()
+  private lateinit var client: OkHttpClient
+  private lateinit var cache: Cache
+  private val NULL_HOSTNAME_VERIFIER =
+    HostnameVerifier { name: String?, session: SSLSession? -> true }
+  private val cookieManager = CookieManager()
+
+  @Before fun setUp() {
+    platform.assumeNotOpenJSSE()
+    platform.assumeNotBouncyCastle()
+    server.protocolNegotiationEnabled = false
+    cache = Cache(File("/cache/"), Int.MAX_VALUE.toLong(), fileSystem)
+    client = clientTestRule.newClientBuilder()
+      .cache(cache)
+      .cookieJar(JavaNetCookieJar(cookieManager))
+      .build()
+  }
+
+  @After fun tearDown() {
+    ResponseCache.setDefault(null)
+    if (this::cache.isInitialized) {
+      cache.delete()
+    }
+  }
+
+  @Test @Throws(IOException::class) fun corruptedCacheEntry() {
+    server.useHttps(handshakeCertificates.sslSocketFactory(), false)
+    server.enqueue(MockResponse()
+      .addHeader("Last-Modified: " + formatDate(-1, TimeUnit.HOURS))
+      .addHeader("Expires: " + formatDate(1, TimeUnit.HOURS))
+      .setBody("ABC"))
+    client = client.newBuilder()
+      .sslSocketFactory(
+        handshakeCertificates.sslSocketFactory(), handshakeCertificates.trustManager)
+      .hostnameVerifier(NULL_HOSTNAME_VERIFIER)
+      .build()
+    val request: Request = Request.Builder().url(server.url("/")).build()
+    val response1: Response = client.newCall(request).execute()
+    val bodySource = response1.body!!.source()
+    Assertions.assertThat(bodySource.readUtf8()).isEqualTo("ABC")
+
+    val metadataFile = fileSystem.files.keys.find { it.name.endsWith(".0") }
+    val metadataBuffer = fileSystem.files[metadataFile]
+
+    val contents = metadataBuffer!!.peek().readUtf8()
+
+    metadataBuffer.writeUtf8(contents.substring(0, contents.length / 4))
```

Comment:
worth a comment? my guess is this is where we are corrupting.. and later we are expecting the request to return the same string anyway... because we are resilient to corrupted entries?

Reply:
Copying the PR notes

> Base for some corruption tests, but currently passing.

I want to try to reproduce every painful cache corruption we can here, make sure the cache always allows for forward progress.

Reply:
sorry I meant a comment to explain this corruption case.

ex. 
```suggestion
    // We expect cached request to serve even when the metadata file is truncated
    metadataBuffer.writeUtf8(contents.substring(0, contents.length / 4))
```

It is a bit more complex I guess to explain how it is able to serve the payload when it is corrupt, though it would be nice to say that if it is short enough to pointer in a sentence.

I've been out a while so have the benefit of low context.. some shared implementation knowledge I'll be able to point out until I too know too much :D